lib/File/Which.pm

```diff
@@ -99,33 +99,86 @@ me.
 our @EXPORT    = 'which';
 our @EXPORT_OK = 'where';
 
-use constant IS_VMS => ($^O eq 'VMS');
-use constant IS_MAC => ($^O eq 'MacOS');
-use constant IS_WIN => ($^O eq 'MSWin32' or $^O eq 'dos' or $^O eq 'os2');
-use constant IS_DOS => IS_WIN();
-use constant IS_CYG => ($^O eq 'cygwin' || $^O eq 'msys');
-
-our $IMPLICIT_CURRENT_DIR = IS_WIN || IS_VMS || IS_MAC;
-
-# For Win32 systems, stores the extensions used for
-# executable files
-# For others, the empty string is used
-# because 'perl' . '' eq 'perl' => easier
-my @PATHEXT = ('');
-if ( IS_WIN ) {
-  # WinNT. PATHEXT might be set on Cygwin, but not used.
-  if ( $ENV{PATHEXT} ) {
-    push @PATHEXT, split /;/, $ENV{PATHEXT};
+sub _get_osname { @_ == 1 && ref $_[0] ? $_[0]->{osname} : $^O }
+
+sub _is_vms { my $osname = &_get_osname; ($osname eq 'VMS'); }
+sub _is_mac { my $osname = &_get_osname; ($osname eq 'MacOS'); }
+sub _is_win { my $osname = &_get_osname; ($osname eq 'MSWin32' or $osname eq 'dos' or $osname eq 'os2'); }
+sub _is_dos { _is_win(@_); }
+sub _is_cyg { my $osname = &_get_osname; ($osname eq 'cygwin' || $osname eq 'msys'); }
+
+sub _default_implicit_current_dir {
+  my $self = shift;
+  $self->_is_win || $self->_is_vms || $self->_is_mac;
+}
+our $IMPLICIT_CURRENT_DIR = do {
+  File::Which->new->_default_implicit_current_dir;
+};
+
+sub new {
+  my ($class, %opts) = @_;
+
+  my $osname = exists $opts{os} ? $opts{os} : $^O;
+
+  my $self = bless {
+    osname => $osname,
+  }, $class;
+
+  $self->{implicit_current_dir} =
+    exists $opts{implicit_current_dir}
+    ? $opts{implicit_current_dir}
+    : $self->_default_implicit_current_dir;
+
+  $self->{PATHEXT} = $self->_default_pathext;
+
+  if( exists $opts{fixed_paths} ) {
+    $self->{fixed_paths} = $opts{fixed_paths};
```

Comment:
I would like to delete from `%opts` and error if there is anything left over, that way we can add new options in the future, and if someone is trying to use those options on an older `File::Which` they will get an error.